File: 2019/Java/d09/wes_computer.java
package d09;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayDeque;
import java.util.ArrayList;

public class wes_computer {
  static final int OP_ADD = 1;    // ADD X1 X2 DEST
  static final int OP_MUL = 2;    // MUL X1 X2 DEST
  static final int OP_STORE = 3;  // STORE <input> DEST
  static final int OP_OUTPUT = 4; // PRINT X1
  static final int OP_JTRUE = 5;  // JTRUE V1 V2 : (if V1!=0 GOTO V2)
  static final int OP_JFALSE = 6; // JFALSE V1 V2: (if V1==0 GOTO V2)
  static final int OP_LT = 7;     // (V1 < V2) -> V3 = 1 ELSE V3 = 0
  static final int OP_EQ = 8;     // (V1 == V2) -> V3 = 1 ELSE V3 = 0
  static final int OP_ADJRB = 9;  // Adjust the relative base...
  static final int OP_END = 99;
  static final String[] OP_NAME = new String[] {"_", "ADD   ", "MUL   ", "STORE ", "OUT   ", "JTRUE ", "JFALSE", "LT    ", "EQ    "};

  static final int MODE_POS = 0;
  static final int MODE_IMM = 1;
  static final int MODE_REL = 2;

  static final int MAX_ARGS = 3;

  public static final byte CONTINUE = 1;
  public static final byte WAIT_INPUT = 2;
  public static final byte HALT = 3;

  private int ip;
  private int relative_base;
  private ArrayDeque<Long> input_stream;
  private ArrayDeque<Long> output_buffer;
  private long[] prog0;
  private ArrayList<Long> prog;
  private boolean verbose;
  private byte status;

  public static wes_computer wc_from_input(String s) throws Exception {
    BufferedReader br = new BufferedReader(new FileReader(s));
    String[] bits = br.readLine().split(",");
    long[] input = new long[bits.length];
    for (int i = 0; i < bits.length; i++) input[i] = Long.parseLong(bits[i]);
    br.close();
    return new wes_computer(input);
  }

  public static long[] toLongs(String msg) {
    long[] lmsg = new long[msg.length()];
    for (int i=0; i<msg.length(); i++) {
      if (msg.charAt(i)=='!') lmsg[i]=10;
      else lmsg[i]=(long) msg.charAt(i);
    } return lmsg;
  }

  public static void printch(long lch) {
    char ch = (char) lch;
    if (ch==10) System.out.println();
    else System.out.print(ch);
  }

  public wes_computer(long[] _prog0, long[] input, boolean _verbose) {
    input_stream = new ArrayDeque<Long>();
    output_buffer = new ArrayDeque<Long>();
    prog = new ArrayList<Long>();
    prog0 = new long[_prog0.length];
    for (int i=0; i<_prog0.length; i++) {
      prog.add(_prog0[i]);
      prog0[i] = _prog0[i];
    }
    add_input(input);
    verbose=_verbose;
    ip=0;
    relative_base=0;
    status = CONTINUE;
  }

  public wes_computer(long[] _prog0) {
    this(_prog0, null, false);
  }

  public long peek(long addr) {
    while (addr >= prog.size()) prog.add(0L);
    return prog.get((int)addr);
  }

  public long peek(long addr, int mode) {
    if (mode == MODE_POS) return peek(peek(addr));
    else if (mode == MODE_IMM) return peek(addr);
    else if (mode == MODE_REL) return peek(peek(addr) + relative_base);
    else { System.out.println("Unknown address mode"); return -1;}
  }

  public wes_computer poke(long addr, long val) {
    while (addr >= prog.size()) prog.add(0L);
    prog.set((int)addr, val);
    return this;
  }

  wes_computer poke(long addr, long val, int mode) {
    addr = peek(addr);
    if (mode==MODE_REL) addr += relative_base;
    if (mode==MODE_IMM) System.out.println("Error - immediate mode assign");
    return poke(addr, val);
  }

  public byte get_status() { return status; }

  public wes_computer add_input(long[] input) {
    if (input!=null) for (int i=0; i<input.length; i++) input_stream.addLast(input[i]);
    return this;
  }

  public int input_queue_size() {
    return input_stream.size();
  }

  public wes_computer add_input(String msg) {
    return add_input(toLongs(msg));
  }

  public wes_computer add_input(long i) {
    input_stream.add(i);
    return this;
  }

  public long read_output() {
    return output_buffer.removeFirst();
  }

  public boolean output_available() {
    return output_buffer.size()>0;
  }

  public wes_computer reset(boolean clear_input, boolean clear_output) {
    prog.clear();
    for (int i=0; i<prog0.length; i++) prog.add(prog0[i]);
    ip=0;
    relative_base=0;
    if (clear_input) input_stream.clear();
    if (clear_output) output_buffer.clear();
    status = CONTINUE;
    return this;
  }

  public wes_computer reset() { return reset(true, true); }

  private int process_math4(int _ip, int op, int[] mode, boolean _verbose) {
    if (_verbose) System.out.println(_ip+": "+OP_NAME[op]+"    "+peek(_ip)+"  "+peek(_ip+1)+"  "+peek(_ip+2)+"  "+peek(_ip+3)+"  rb = "+relative_base);
         if (op == OP_ADD) poke(_ip+3, peek(_ip+1, mode[0]) + peek(_ip+2, mode[1]), mode[2]);
    else if (op == OP_MUL) poke(_ip+3, peek(_ip+1, mode[0]) * peek(_ip+2, mode[1]), mode[2]);
    else if (op == OP_LT)  poke(_ip+3, (peek(_ip+1, mode[0]) < peek(_ip+2, mode[1])?1:0), mode[2]);
    else if (op == OP_EQ)  poke(_ip+3, (peek(_ip+1, mode[0]) == peek(_ip+2, mode[1])?1:0), mode[2]);
    return ip+4;
  }

  private int process_jump(int _ip, int op, int[] mode, boolean _verbose) {
    if (_verbose) System.out.println(_ip+": JTRUE  "+peek(_ip)+"  "+peek(_ip+1)+"   "+peek(_ip+2)+"  rb = "+relative_base);
    long val = peek(_ip+1, mode[0]);
    long dest = peek(_ip+2, mode[1]);

    if (op == OP_JTRUE) return (int) ((val != 0) ? dest : _ip+3);
    else /*if (op == OP_JFALSE)*/ return (int) ((val ==0) ? dest : _ip+3);
  }

  public wes_computer run() {
    status = CONTINUE;
    while(status == CONTINUE) status = exec();
    return this;
  }

  byte exec() {
    int[] mode = new int[MAX_ARGS];
    int opcode = prog.get(ip).intValue();
    int op = opcode % 100;
    opcode /= 100;
    for (int j=0; j<MAX_ARGS; j++) {
      mode[j] = opcode % 10;
      opcode /= 10;
    }

    if ((op == OP_ADD) || (op == OP_MUL) || (op == OP_EQ) || (op == OP_LT)) {
      ip = process_math4(ip, op, mode, verbose);

    } else if ((op == OP_JTRUE) || (op == OP_JFALSE)) {
      ip = process_jump(ip, op, mode, verbose);

    } else if (op == OP_STORE) {
      if (verbose) System.out.println(ip+": STORE    "+peek(ip)+"  "+peek(ip+1)+"  input stream = "+input_stream.getFirst()+"  rb = "+relative_base);
      if (input_stream.size() == 0) return WAIT_INPUT;
      long val = input_stream.removeFirst();
      poke(ip + 1, val, mode[0]);
      ip+=2;

    } else if (op == OP_OUTPUT) {
      if (verbose) System.out.println(ip+": OUTPUT    "+peek(ip)+"  "+peek(ip+1)+"  rb = "+relative_base);
      long val = peek(ip+1, mode[0]);
      output_buffer.addLast(val);
      if (verbose) System.out.println("INTO OUTPUT BUFFER: "+val);
      ip+=2;

    } else if (op == OP_ADJRB) {
      if (verbose) System.out.println(ip+": ADJ_RB    "+peek(ip)+"   "+peek(ip+1)+"   old_rb = "+relative_base);
      long val1 = peek(ip+1, mode[0]);
      relative_base += val1;
      ip+=2;

    } else if (op == OP_END) return HALT;

    return CONTINUE;
  }
}

File: 2019/R/wes_computer.R
library(R6)

ZVec <- R6Class(
  classname = "ZVec",
  private = list(
    mainvec = NULL,
    zeroelement = NULL
  ),
  
  public = list(
    initialize = function(content) {
      private$zeroelement <- content[1]
      private$mainvec <- content[2:length(content)]
    },
    
    print = function(...) {
      print(c(private$zeroelement, private$mainvec))
      invisible(self)
    },
    
    copy = function() {
      ZVec$new(c(private$zeroelement, private$mainvec))
    },
    
    at = function(x) {
      if (x == 0) {
        return(private$zeroelement)
      } else {
        return(private$mainvec[x])
      }
    },
    
    set = function(addr, val) {
      self$ensure_space(addr)
      if (addr == 0) {
        private$zeroelement <- val
      } else {
        private$mainvec[addr] <- val
      }
    },
    
    size = function() {
      length(private$zeroelement) + length(private$mainvec)
    },
    
    ensure_space = function(len) {
      extra <- 1 + (len - self$size())
      if (extra > 0) {
        private$mainvec <- c(private$mainvec, rep(0, extra))
      }
      invisible(self)
    }
  )
)

IntComputer <- R6Class(
  classname = "IntComputer", 
  
  private = list(
    ip = 0L,
    relative_base = 0L,
    input_stream = NULL,
    output_buffer = NULL,
    program = NULL,
    backup_program = NULL,
    status = 0L,
    
    OP_ADD    = 1L,   # ADD X1 X2 DEST
    OP_MUL    = 2L,   # MUL X1 X2 DEST
    OP_STORE  = 3L,   # STORE <input> DEST
    OP_OUTPUT = 4L,   # PRINT X1 to <output>
    OP_JTRUE  = 5L,   # JTRUE V1 V2 : (if V1!=0 GOTO V2)
    OP_JFALSE = 6L,   # JFALSE V1 V2: (if V1==0 GOTO V2)
    OP_LT     = 7L,   # (V1 < V2) -> V3 = 1 ELSE V3 = 0
    OP_EQ     = 8L,   # (V1 == V2) -> V3 = 1 ELSE V3 = 0
    OP_ADJRB  = 9L,   # Adjust the relative base...
    OP_END    = 99L,
    OP_NAME   = c("ADD", "MUL", "STO", "OUT", "JT", "JF", "LT", "EQ", "ARB"),
    
    MODE_POS = 0L,
    MODE_IMM = 1L,
    MODE_REL = 2L,
    
    peek2 = function(addr) {
      private$program$ensure_space(addr)
      return(private$program$at(addr))
    },
    
    poke2 = function(addr, value) {
      if (is.na(value)) message("POKE NA")
      private$program$set(addr, value)
      invisible(self)
    },
    
    process_math4 = function(ip, op, mode, verbose=FASLE) {
      thing1 <- self$peek(ip + 1, mode[1])
      thing2 <- self$peek(ip + 2, mode[2])
      dest <- ip + 3
      
      if (verbose)
        message(sprintf("%d:\t%s\t%d\t%d\t%d\t%d\tmode = %d%d%d\trb = %d",
                        private$ip, private$OP_NAME[op], 
                        self$peek(private$ip), self$peek(private$ip + 1), self$peek(private$ip + 2),
                        self$peek(private$ip + 3), mode[1], mode[2], mode[3],
                        private$relative_base)) 
      
           if (op == private$OP_ADD)  self$poke(dest, thing1 + thing2, mode[3])
      else if (op == private$OP_MUL)  self$poke(dest, thing1 * thing2, mode[3])
      else if (op == private$OP_LT)   self$poke(dest, as.numeric(thing1 < thing2), mode[3])
      else if (op == private$OP_EQ)   self$poke(dest, as.numeric(thing1 == thing2), mode[3])
      
      ip + 4 
    },
    
    process_jump = function(ip, op, mode, verbose = FALSE) {
      if (verbose) 
        message(sprintf("%d:\t%s\t%d\t%d\t%d\t\tmode = %d%d\trb = %d",
                        ip, private$OP_NAME[op], self$peek(ip), self$peek(ip+1),
                        self$peek(ip+2), mode[1], mode[2], 
                        private$relative_base))
      val <- self$peek(private$ip + 1, mode[1])
      dest <- self$peek(private$ip + 2, mode[2])
   
      if (op == private$OP_JTRUE) { 
        if (val !=0) return(dest)
        else return(private$ip + 3)
      
      } else if (op == private$OP_JFALSE) {
        if (val == 0) return(dest)
        else return(private$ip + 3)
      }
      
      dest
    },
    
    exec = function(verbose) {
      mode <- c(0,0,0)
      opcode <- as.numeric(private$program$at(private$ip))
      op <- opcode %% 100
      opcode <- as.integer(opcode / 100)
      mode[1] <- opcode %% 10
      opcode <- as.integer(opcode / 10)
      mode[2] <- opcode %% 10
      opcode <- as.integer(opcode / 10)
      mode[3] <- opcode %% 10

      if (op %in% c(private$OP_ADD, private$OP_MUL,
                    private$OP_EQ,  private$OP_LT)) {
        private$ip <- private$process_math4(private$ip, op, mode, verbose)
      
      } else if (op %in% c(private$OP_JTRUE, private$OP_JFALSE)) {
        private$ip <- private$process_jump(private$ip, op, mode, verbose)
      
      } else if (op == private$OP_STORE) {
        if (verbose)
          message(sprintf("%d:\t%s\t%d\t%d\t\t\tmode = %d\trb = %d",
                          private$ip, private$OP_NAME[op], 
                          self$peek(private$ip), self$peek(private$ip + 1), mode[1], 
                          private$relative_base)) 
        if (length(private$input_stream) == 0) {
          return(self$WAIT_INPUT)
        }
        
        val <- private$input_stream[1]
        private$input_stream <- private$input_stream[-1]
        self$poke(private$ip + 1, val, mode[1])
        private$ip <- private$ip + 2
      
      } else if (op == private$OP_OUTPUT) {
        if (verbose)
          message(sprintf("%d:\t%s\t%d\t%d\t\t\tmode = %d\trb = %d",
                          private$ip, private$OP_NAME[op], 
                          self$peek(private$ip), self$peek(private$ip+1), 
                          mode[1], private$relative_base)) 
        val <- self$peek(private$ip + 1, mode[1])
        private$output_buffer <- c(private$output_buffer, val)
        private$ip <- private$ip + 2
        
      } else if (op == private$OP_ADJRB) {
        if (verbose)
          message(sprintf("%d:\t%s\t%d\t%d\tmode = %d\trb = %d",
                          private$ip, private$OP_NAME[op], 
                          self$peek(private$ip), self$peek(private$ip + 1), 
                          mode[1], private$relative_base)) 
        
        val <- self$peek(private$ip + 1, mode[1])
        private$relative_base <- private$relative_base + val
        private$ip <- private$ip + 2
      
      } else if (op == private$OP_END) {
        if (verbose)
          message(sprintf("%d:\tHALT\t%d", private$ip, op))
        return(self$HALT)
      }
      self$CONTINUE
    }

  ),

  public = list(
    CONTINUE = 1L,
    WAIT_INPUT = 2L,
    HALT = 3L,
    
    initialize = function(file = NULL, program = NULL) {
      if ((is.null(file) + is.null(program)) != 1) {
        stop("Either file, or program must be non-null");
      }
      if (!is.null(file))
        private$program <- ZVec$new(as.numeric(unlist(strsplit(readLines(file), ","))))
      else 
        private$program <- ZVec$new(program)
      private$backup_program <- private$program$copy()
      self$reset()
      invisible(self)
    },
    
    peek = function(address, mode = private$MODE_IMM) {
      if (mode == private$MODE_POS) return(private$peek2(private$peek2(address)))
      else if (mode == private$MODE_IMM) return(private$peek2(address))
      else if (mode == private$MODE_REL) return(private$peek2(private$peek2(address) + 
                                                              private$relative_base))
    },
  
    poke = function(address, value, mode = NULL) {
      if (!is.null(mode)) {
        address <- private$peek2(address) 
        if (mode == private$MODE_REL) address <- address + private$relative_base
      }
      return(private$poke2(address, value))
    },

    get_status = function() {
      private$status
    },
    
    add_input = function(num) {
      private$input_stream <- c(private$input_stream, num)
      invisible(self)
    },
    
    input_size = function() {
      length(private$input_stream)
    },
    
    read_output = function() {
      val <- private$output_buffer[1]
      private$output_buffer <- private$output_buffer[-1]
      val
    },
    
    output_available = function() {
      length(private$output_buffer) > 0
    },
    
    reset = function(clear_input = TRUE, clear_output = TRUE) {
      private$program <- private$backup_program$copy()
      private$ip <- 0
      private$relative_base <- 0
      if (clear_input) private$input_stream <- NULL
      if (clear_output) private$output_buffer <- NULL
      private$status <- self$CONTINUE
      invisible(self)
    },
    
    run = function(verbose = FALSE) {
      private$status = self$CONTINUE
      while (self$get_status() == self$CONTINUE) {
        private$status <- private$exec(verbose)
      }
      invisible(self)
    },
    
    dump = function() {
      print(private$program)
    }
  )
)
